### PR TITLE
Feature/feature pet meal

### PR DESCRIPTION
**Pull request by marclll**

### CODE REVIEW

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        petName = "Canpeque2";
+        date = "2019-02-13T10:30:00";
+        date2 = "2021-02-13T10:30:00";
+        field = "mealName";
+        value = "Asparagus with ketchup";
+        urlBase = "/pet/meal";
+    }
+
+    @Test
+    public void createMealShouldReturnStatusOk() throws Exception {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:73]: Codacy Issue found: [Throwing 'Exception' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/service/MealServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962&bid=17353375)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Test
+    public void getMealDataShouldReturnMealEntityAndStatusOk() throws Exception {
+        willReturn(mealEntity).given(service).getMealData(anyString(), anyString(), anyString());
+        mockMvc.perform(get(urlBase + "/" + owner + "/" + petName + "/" + date))
+            .andExpect(status().isOk());
+    }
+
+    @Test
+    public void getAllMealDataShouldReturnMealEntityListAndStatusOk() throws Exception {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:104]: Codacy Issue found: [Throwing 'Exception' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    private static String value;
+    private static String urlBase;
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private MealService service;
+
+    @BeforeAll
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:52]: Codacy Issue found: [Method length is 18 lines (max allowed is 15).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/service/MealServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class MealServiceTest {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/service/MealServiceTest.java:26]: Codacy Issue found: [This class has too many methods, consider refactoring it.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    @Test
+    public void getAllMealsBetweenShouldReturnMealEntityListAndStatusOk() throws Exception {
+        willReturn(mealEntityList).given(service).getAllMealsBetween(anyString(), anyString(), anyString(),
+            anyString());
+        mockMvc.perform(get(urlBase + "/" + owner + "/" + petName + "/between/" + date + "/" + date2))
+            .andExpect(status().isOk());
+    }
+
+    @Test
+    public void getMealFieldShouldReturnFieldValueAndStatusOk() throws Exception {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:119]: Codacy Issue found: [Throwing 'Exception' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/service/MealServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                isA(String.class));
+            service.getMealData(owner, petName, date);
+        }, "Should return an exception when retrieving a meal fails");
+    }
+
+    @Test
+    public void shouldReturnMealEntityListWhenGetSetOfMealsRetrieved() throws DatabaseAccessException {
+        when(mealDao.getAllMealData(owner, petName)).thenReturn(mealList);
+        List<Map<String, Object>> list = service.getAllMealData(owner, petName);
+        assertSame(mealList, list, "Should return a list of meal entities");
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/service/MealServiceTest.java:100]: Codacy Issue found: [The String "Should return a list of meal entities" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        date2 = "2021-02-13T10:30:00";
+        field = "mealName";
+        value = "Asparagus with ketchup";
+        urlBase = "/pet/meal";
+    }
+
+    @Test
+    public void createMealShouldReturnStatusOk() throws Exception {
+        willDoNothing().given(service).createMeal(anyString(), anyString(), anyString(), isA(MealEntity.class));
+        mockMvc.perform(post(urlBase + "/" + owner + "/" + petName + "/" + date)
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:75]: Codacy Issue found: [The String "/" appears 24 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+            .contentType(MediaType.APPLICATION_JSON)
+            .content(jsonMealEntity))
+            .andExpect(status().isOk());
+    }
+
+    @Test
+    public void deleteByDateShouldReturnStatusOk() throws Exception {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:82]: Codacy Issue found: [Throwing 'Exception' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        willDoNothing().given(service).deleteByDate(anyString(), anyString(), anyString());
+        mockMvc.perform(delete(urlBase + "/" + owner + "/" + petName + "/" + date)
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk());
+    }
+
+    @Test
+    public void deleteAllMealsShouldReturnStatusOk() throws Exception {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:90]: Codacy Issue found: [Throwing 'Exception' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private MealService service;
+
+    @BeforeAll
+    public static void setUp() {
+        jsonMealEntity = "{\n"
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:54]: Codacy Issue found: [The String "{\n" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Test
+    public void deleteAllMealsShouldReturnStatusOk() throws Exception {
+        willDoNothing().given(service).deleteAllMeals(anyString(), anyString());
+        mockMvc.perform(delete(urlBase + "/" + owner))
+            .andExpect(status().isOk());
+    }
+
+    @Test
+    public void getMealDataShouldReturnMealEntityAndStatusOk() throws Exception {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestControllerTest.java:97]: Codacy Issue found: [Throwing 'Exception' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962&bid=17353375)

File: src/main/java/org/pesmypetcare/webservice/dao/MealDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        List<Map<String, Object>> externalList = new ArrayList<>();
+        try {
+            getMealsBetweenDatesFromDatabase(initialDate, finalDate, mealsRef, externalList);
+        } catch (InterruptedException | ExecutionException e) {
+            throw new DatabaseAccessException(DELFAIL_KEY, e.getMessage());
+        }
+        return externalList;
+    }
+
+    private void getMealsBetweenDatesFromDatabase(String initialDate, String finalDate, CollectionReference mealsRef,
```
Comment (AlbertPG) [inline src/main/java/org/pesmypetcare/webservice/dao/MealDaoImpl.java:134]: You should comment this method

File: src/main/java/org/pesmypetcare/webservice/dao/MealDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Override
+    public void updateMealField(String owner, String petName, String date, String field, Object value) {
+        CollectionReference mealsRef = getMealsRef(owner, petName);
+        DocumentReference mealRef = mealsRef.document(date);
+        mealRef.update(field, value);
+    }
+
+    public CollectionReference getMealsRef(String owner, String petName) {
```
Comment (AlbertPG) [inline src/main/java/org/pesmypetcare/webservice/dao/MealDaoImpl.java:174]: You should comment this

**Review by AlbertPG — COMMENTED** [review]
You should comment this methods

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962&bid=17353375)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962&bid=17353375)

File: src/main/java/org/pesmypetcare/webservice/dao/MealDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        try {
+            getAllMealOfAPetFromDatabase(mealsRef, externalList);
+        } catch (InterruptedException | ExecutionException e) {
+            throw new DatabaseAccessException(DELFAIL_KEY, e.getMessage());
+        }
+        return externalList;
+    }
+
+    /**
+     * Gets all the meals of the collection and puts them in the externalList
```
Comment (xavier-campos-diaz) [inline src/main/java/org/pesmypetcare/webservice/dao/MealDaoImpl.java]: Missing end period

**Review by AlbertPG — APPROVED** [review]
Okey

**Review by xavier-campos-diaz — APPROVED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5301962&bid=17353375)

**Pull request merged**